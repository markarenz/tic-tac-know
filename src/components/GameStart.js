import React, { useState } from 'react';
import { Transition } from '@tailwindui/react';
import { sleep } from '../helpers/gameLogic';
import { sides, coinSides } from '../constants';
import imgHeads from '../img/heads.png';
import imgTails from '../img/tails.png';
import PropTypes from 'prop-types';
const GameStart = ({ show, handleStartClick }) => {
  const [isExisting, setIsExiting] = useState(false);
  const [rndIndex, setRndIndex] = useState(null);
  if (rndIndex === null) {
    setRndIndex(Math.floor(Math.random() * sides.length));
  }
  const coinFlip = coinSides[rndIndex];
  const coinFlipWhoseTurn = sides[rndIndex];
  const imgSrcArr = [imgHeads, imgTails];
  const handleModalStartClick = async () => {
    setIsExiting(true);
    await sleep(500);
    setIsExiting(false);
    handleStartClick(coinFlipWhoseTurn);
  };
  if (!coinFlip) {
    return null;
  }
  return (
    <Transition
      as="div"
      show={show && !isExisting}
      enter="transition-all duration-500"
      enterFrom="opacity-0 -translate-x-32"
      enterTo="opacity-100 translate-x-0"
      leave="transition-all duration-500"
      leaveFrom="opacity-100 translate-x-0"
      leaveTo="opacity-0 translate-x-32"
      className="absolute left-0 top-0 w-full h-full flex items-center justify-center"
    >
      <div className="bg-purple-900 text-white border-2 border-purple-400 rounded-lg p-4 w-full max-w-lg drop-shadow-xl">
        <div className="grid grid-cols-4">
          <div className="pr-2">
            <div className="h-full flex justify-center align-center">
              <div>
                <img src={imgSrcArr[rndIndex]} alt={coinFlip} />
              </div>
            </div>
          </div>
          <div className="col-span-3">
            <div className="flex justify-center items-center h-full">
              <div>
                <h2 className="text-3xl text-center">
                  <i>The coin toss came up {coinFlip}.</i>
                </h2>
                <h3 className="text-xl text-center">
                  {coinFlipWhoseTurn.toUpperCase()} goes first this time.
                </h3>
              </div>
            </div>
          </div>
        </div>
        <div className="text-right">
          <button
            className="uppercase bg-transparent text-white border-2 border-white hover:bg-white hover:text-purple-900 rounded-md py-2 px-4 transition-all hover:scale-110 duration-300"
            onClick={handleModalStartClick}
          >
            Begin
          </button>
        </div>
      </div>
    </Transition>
  );
};

GameStart.defaultPropTypes = {
  show: false,
};

GameStart.propTypes = {
  show: PropTypes.bool.isRequired,
  handleStartClick: PropTypes.func.isRequired,
};

export default GameStart;
