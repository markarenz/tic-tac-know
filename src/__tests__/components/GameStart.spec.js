import { render, fireEvent, screen, waitFor } from '@testing-library/react';
// import { act } from 'react-dom/test-utils';
import { mockRandom } from 'jest-mock-random';
import React, { useState } from 'react';
import { act } from 'react-dom/test-utils';
import GameStart from '../../components/GameStart';

const mocks = {
  handleStartClick: jest.fn(),
};

beforeEach(() => {
  mockRandom(0);
});

describe('GameStart', () => {
  it('matches snapshot', () => {
    const container = render(<GameStart show={true} handleStartClick={mocks.handleStartClick} />);
    expect(container).toMatchSnapshot();
  });
  it('matches snapshot show = false', () => {
    const container = render(<GameStart show={false} handleStartClick={mocks.handleStartClick} />);
    expect(container).toMatchSnapshot();
  });

  it('reacts when start is clicked', async () => {
    // jest.useFakeTimers();
    render(<GameStart show={true} handleStartClick={mocks.handleStartClick} />);
    fireEvent(
      screen.getByTestId('gameStart-btn-start'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    );
    await waitFor(() => {
      expect(mocks.handleStartClick).toHaveBeenCalled();
    });
  });
});
