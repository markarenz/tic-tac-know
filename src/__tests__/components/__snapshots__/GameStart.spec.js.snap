// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameStart matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-testid="gamestart"
      >
        <div
          class=""
          data-testid="transition-wrap"
        >
          <div
            class="absolute left-0 top-0 w-full h-screen bg-black/60"
          />
        </div>
        <div
          class="absolute left-0 top-0 w-full h-screen flex items-center justify-center"
          data-testid="transition-wrap"
        >
          <div
            class="bg-purple-900 text-white border-2 border-purple-400 rounded-lg p-4 w-full max-w-lg drop-shadow-xl"
          >
            <div
              class="grid grid-cols-4"
            >
              <div
                class="pr-2"
              >
                <div
                  class="h-full flex justify-center align-center"
                >
                  <div>
                    <img
                      alt="heads"
                      src="heads.png"
                    />
                  </div>
                </div>
              </div>
              <div
                class="col-span-3"
              >
                <div
                  class="flex justify-center items-center h-full"
                >
                  <div>
                    <h2
                      class="text-3xl text-center"
                    >
                      <i>
                        The coin toss came up 
                        heads
                        .
                      </i>
                    </h2>
                    <h3
                      class="text-xl text-center"
                    >
                      PLAYER
                       goes first this time.
                    </h3>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="text-right"
            >
              <button
                class="uppercase bg-transparent text-white border-2 border-white hover:bg-white hover:text-purple-900 rounded-md py-2 px-4 transition-all hover:scale-110 duration-300"
                data-testid="gameStart-btn-start"
              >
                Begin
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-testid="gamestart"
    >
      <div
        class=""
        data-testid="transition-wrap"
      >
        <div
          class="absolute left-0 top-0 w-full h-screen bg-black/60"
        />
      </div>
      <div
        class="absolute left-0 top-0 w-full h-screen flex items-center justify-center"
        data-testid="transition-wrap"
      >
        <div
          class="bg-purple-900 text-white border-2 border-purple-400 rounded-lg p-4 w-full max-w-lg drop-shadow-xl"
        >
          <div
            class="grid grid-cols-4"
          >
            <div
              class="pr-2"
            >
              <div
                class="h-full flex justify-center align-center"
              >
                <div>
                  <img
                    alt="heads"
                    src="heads.png"
                  />
                </div>
              </div>
            </div>
            <div
              class="col-span-3"
            >
              <div
                class="flex justify-center items-center h-full"
              >
                <div>
                  <h2
                    class="text-3xl text-center"
                  >
                    <i>
                      The coin toss came up 
                      heads
                      .
                    </i>
                  </h2>
                  <h3
                    class="text-xl text-center"
                  >
                    PLAYER
                     goes first this time.
                  </h3>
                </div>
              </div>
            </div>
          </div>
          <div
            class="text-right"
          >
            <button
              class="uppercase bg-transparent text-white border-2 border-white hover:bg-white hover:text-purple-900 rounded-md py-2 px-4 transition-all hover:scale-110 duration-300"
              data-testid="gameStart-btn-start"
            >
              Begin
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`GameStart matches snapshot show = false 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-testid="gamestart"
      />
    </div>
  </body>,
  "container": <div>
    <div
      data-testid="gamestart"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
